rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala

```diff
@@ -0,0 +1,173 @@
+                     syntheticSamples: Int,
+                     thin: Int)(implicit rng: RNG): (Int, Double, Double) = {
+    val trueValues = priorGenerator.get(rng, emptyEvaluator)
+    implicit val trueEval = new Evaluator(priorParams.zip(trueValues).toMap)
+    val trueOutput = posterior.map(_._2).get
+
+    val syntheticValues =
+      posterior.map(_._1.generator.repeat(syntheticSamples)).get
+    val model = posterior.flatMap {
+      case (d, r) =>
+        d.fit(syntheticValues).map { _ =>
+          r
+        }
+    }
+    val (samples, diag) = model.sampleWithDiagnostics(sampler,
+                                                      Chains,
+                                                      warmupIterations,
+                                                      (Samples / Chains) * thin,
+                                                      thin)
+
+    val maxRHat = diag.map(_.rHat).max
+    val minEffectiveSampleSize = diag.map(_.effectiveSampleSize).min
+    val rawRank = samples.tail.count { n =>
+      n < trueOutput
+    }
+    (rawRank, maxRHat, minEffectiveSampleSize)
+  }
+
+  case class Run(sampler: Sampler,
+                 warmupIterations: Int,
+                 syntheticSamples: Int,
+                 thin: Int,
+                 rHat: Double)(implicit rng: RNG) {
+
+    def animate(logBins: Int): Stream[Int] = {
+      val t0 = System.currentTimeMillis
+      val stream = simulate(logBins)
+      val bins = 1 << logBins
+      val reps = bins * RepsPerBin
+
+      val lower = binomialQuantile(0.005, reps, 1.0 / bins)
+      val upper = binomialQuantile(0.995, reps, 1.0 / bins)
+      println("\n" * (bins + 1))
+      1.to(reps).foreach { i =>
+        val list = stream.take(i).toList
+        val timeTaken = System.currentTimeMillis - t0
+        val timeRemaining = timeTaken * (reps - i) / i
+        plot(list, bins, i, reps, lower, upper, timeRemaining)
+      }
+      stream
+    }
+
+    def simulate(logBins: Int): Stream[Int] = {
+      require(logBins > 0)
+      val bins = 1 << logBins
+      require(bins <= Samples)
+
+      val reps = bins * RepsPerBin
+      rankStream(bins, reps)
+    }
+
+    private def rankStream(bins: Int, remaining: Int): Stream[Int] =
+      if (remaining == 0)
+        Stream.empty
+      else {
+        val (rawRank, _, _) =
+          sample(sampler, warmupIterations, syntheticSamples, thin)
+        val rank = (rawRank * bins) / Samples
+        rank #:: rankStream(bins, remaining - 1)
+      }
+
+    private def plot(list: List[Int],
+                     bins: Int,
+                     rep: Int,
+                     reps: Int,
+                     lower: Int,
+                     upper: Int,
+                     millisRemaining: Long): Unit = {
+      println("\u001b[1000D") //move left
+      println(s"\u001b[${bins + 3}A") //move up
+      val remaining = formatMillis(millisRemaining)
+      println(s"Repetition $rep/$reps. Estimated time remaining: $remaining")
```

Comment:
This dynamic plot is really great! Could we add a line here explaining that the `[` and `]` bracket the expected 0.5-99.5th percentiles in each bin?

Reply:
Oh, and you mentioned printing rHat, but I don't see it. Maybe print rHat and any other interesting diagnostics?

Reply:
Will do.